### PR TITLE
DDP-8576: Picklist option exclusivity changes in PATIENT_SURVEY and ABOUT_HEALTHY

### PR DESCRIPTION
**Pull request by bskinner**
DDP-8576

## Context

This PR adds a patch with 2 changes to existing questions in the `PATIENT_SURVEY` and `ABOUT_HEALTHY` surveys. The changes are as below:
|Survey|Question|Option|Notes|
|:----:|:-:|:-:|:-:|
|ABOUT_HEALTHY|IF_YOU_HAVE_HAD_ARRHYTHMIA|DID_NOT_REQUIRED|Set option to be `exclusive`|
|PATIENT_SURVEY|IF_YOU_HAVE_HAD_ARRHYTHMIA_COMPLICATIONS|DO_NOT_REQUIRE|Set option to be `exclusive`|

This change does not require that a new revision be created for these modifications.

## Applying the changes
The changes in this PR can be applied using the `--run-task` or `--patch` options to `studybuilder-cli` (replacing `path/to/my/application.conf` with the path to your actual configuration)
```
java \
    -Dconfig.file=path/to/my/application.conf \
    -cp studybuilder-cli/target/StudyBuilder.jar \
    org.broadinstitute.ddp.studybuilder.StudyBuilderCli \
      --vars "studybuilder-cli/output-config/vars.conf" \
      --substitutions "studybuilder-cli/studies/singular/substitutions.conf" \
      --process-translations "PROCESS_IGNORE_TEMPLATES_WITH_TRANSLATIONS" \
      --translations-to-db-json \
      --run-task "SingularAboutHealthySelectLogic" \
      "studybuilder-cli/studies/singular/study.conf"
```

### CODE REVIEW

File: pepper-apis/dss-core/src/main/java/org/broadinstitute/ddp/db/dto/PicklistOptionDto.java [hunk 1]
```diff
@@ -1,17 +1,17 @@
...
 import java.util.HashSet;
 import java.util.List;
 import java.util.Set;
+import javax.annotation.Nullable;
+
+import lombok.AllArgsConstructor;
+import lombok.Data;
+import lombok.experimental.Accessors;
+import org.jdbi.v3.core.mapper.reflect.ColumnName;
+import org.jdbi.v3.core.mapper.reflect.JdbiConstructor;
```
Comment (bskinner) [inline pepper-apis/dss-core/src/main/java/org/broadinstitute/ddp/db/dto/PicklistOptionDto.java:14]: No major changes here, just quieting Checkstyle

File: pepper-apis/studybuilder-cli/src/main/java/org/broadinstitute/ddp/studybuilder/task/SingularAboutHealthySelectLogic.java [hunk 2]
```diff
@@ -0,0 +1,220 @@
...
+
+/**
+ * Task for updating the IF_YOU_HAVE_HAD_ARRHYTHMIA_COMPLICATIONS (PATIENT_SURVEY)
+ * and IF_YOU_HAVE_HAD_ARRHYTHMIA (ABOUT_HEALTHY) questions to mark the DO_NOT_REQUIRE
+ * and DID_NOT_REQUIRED options (respectively) as exclusive.
+ * 
+ * <p>See https://broadinstitute.atlassian.net/browse/DDP-8576
+ */
+@Slf4j
+public class SingularAboutHealthySelectLogic implements CustomTask {
```
Comment (bskinner) [inline pepper-apis/studybuilder-cli/src/main/java/org/broadinstitute/ddp/studybuilder/task/SingularAboutHealthySelectLogic.java]: Suggestions for names would be welcomed here- this name is somewhat misleading as this is altering both the `ABOUT_HEALTHY` and `PATIENT_SURVEY` activities.

File: pepper-apis/studybuilder-cli/src/main/java/org/broadinstitute/ddp/studybuilder/task/SingularAboutHealthySelectLogic.java [hunk 3]
```diff
@@ -0,0 +1,220 @@
...
+                    .filter((activityDef) -> activityDef.getActivityType() == ActivityType.FORMS)
+                    .map(FormActivityDef.class::cast)
+                    .findFirst()
+                    .orElseThrow(() -> {
+                        final var message = String.format("Unable to find the activity [%s:s] with code '%s' in study %s]",
+                                activityCode,
+                                versionTag,
+                                ActivityType.FORMS,
+                                studyGuid);
+                        throw new DDPException(message);
```
Comment (bskinner) [inline pepper-apis/studybuilder-cli/src/main/java/org/broadinstitute/ddp/studybuilder/task/SingularAboutHealthySelectLogic.java]: This should be `return new`, not `throw new`

File: pepper-apis/studybuilder-cli/src/main/java/org/broadinstitute/ddp/studybuilder/task/SingularAboutHealthySelectLogic.java [hunk 4]
```diff
@@ -0,0 +1,220 @@
...
+        return Optional.ofNullable(activityDef.getQuestionByStableId(stableId)).stream()
+                .filter((questionDef) -> questionDef.getQuestionType() == QuestionType.PICKLIST)
+                .map(PicklistQuestionDef.class::cast)
+                .findFirst()
+                .orElseThrow(() -> {
+                    final var message = String.format("Unable to find a question with stable id '%s' and type '%s' in activity %s]",
+                            stableId,
+                            QuestionType.PICKLIST,
+                            activityDef.getActivityCode());
+                    throw new DDPException(message);
```
Comment (bskinner) [inline pepper-apis/studybuilder-cli/src/main/java/org/broadinstitute/ddp/studybuilder/task/SingularAboutHealthySelectLogic.java]: This should be `return new`, not `throw new`

File: pepper-apis/studybuilder-cli/src/main/java/org/broadinstitute/ddp/studybuilder/task/SingularAboutHealthySelectLogic.java [hunk 5]
```diff
@@ -0,0 +1,220 @@
...
+
+    @Override
+    public void init(Path configPath, Config studyConfig, Config varsConfig) {
+        var studyRoot = configPath.getParent();
+        var patchConfPath = Paths.get(PATCH_PATH, PATCH_CONF_NAME);
+        var patchPath = studyRoot.resolve(patchConfPath);
+
+        final Config patchConfig;
+        try {
+            patchConfig = ConfigFactory.parseFile(patchPath.toFile()).resolveWith(varsConfig);
```
Comment (bskinner) [inline pepper-apis/studybuilder-cli/src/main/java/org/broadinstitute/ddp/studybuilder/task/SingularAboutHealthySelectLogic.java]: `ConfigFactor.parseFile(File)` does not error out of the target file does not exist. Explicitly check for the patch configuration file & bail early if it's not there.

File: pepper-apis/studybuilder-cli/src/main/java/org/broadinstitute/ddp/studybuilder/task/SingularAboutHealthySelectLogic.java [hunk 6]
```diff
@@ -0,0 +1,220 @@
...
+
+    private JdbiUmbrellaStudy studySql;
+    private JdbiActivity activitySql;
+    private JdbiActivityVersion activityVersionSql;
+    private FormActivityDao activityDao;
+    private JdbiPicklistOption picklistOptionSql;
+
+    @Override
+    public void init(Path configPath, Config studyConfig, Config varsConfig) {
+        var studyRoot = configPath.getParent();
```
Comment (bskinner) [inline pepper-apis/studybuilder-cli/src/main/java/org/broadinstitute/ddp/studybuilder/task/SingularAboutHealthySelectLogic.java]: This intermediate value isn't necessary- roll this into `patchPath`

File: pepper-apis/studybuilder-cli/src/main/java/org/broadinstitute/ddp/studybuilder/task/SingularAboutHealthySelectLogic.java [hunk 7]
```diff
@@ -0,0 +1,220 @@
...
+        var targetStudy = patchConfig.getString(Keys.ChangeSet.STUDY);
+        if (!targetStudy.equals(studyGuid)) {
+            throw new DDPException(String.format("Patch %s targets study '%s', but is being run against '%s'",
+                    this.getClass().getSimpleName(),
+                    targetStudy,
+                    studyGuid));
+        }
+
+        this.patchConfig = patchConfig;
+        this.studyConfig = studyConfig;
```
Comment (bskinner) [inline pepper-apis/studybuilder-cli/src/main/java/org/broadinstitute/ddp/studybuilder/task/SingularAboutHealthySelectLogic.java]: `studyConfig` isn't used anywhere outside of this method- the property can be removed.

File: pepper-apis/studybuilder-cli/src/main/java/org/broadinstitute/ddp/studybuilder/task/SingularAboutHealthySelectLogic.java (hunk 2 above)
Comment (MocanaAtBroad) [reply pepper-apis/studybuilder-cli/src/main/java/org/broadinstitute/ddp/studybuilder/task/SingularAboutHealthySelectLogic.java]: How about SingularMakeNoMedicalTreatmentOptionExclusive? (Though I am going to drink very large beer in celebration when we get rid of the patch thing and these java "tasks" and weird class names are an unpleasant memory!)

Comment (bskinner) [reply pepper-apis/studybuilder-cli/src/main/java/org/broadinstitute/ddp/studybuilder/task/SingularAboutHealthySelectLogic.java]: Seconded 💯.

That name works for me.

**Comment by bskinner** [comment]
Tests are currently failing due to the known flaky `org.broadinstitute.ddp.service.userdelete.UserDeleteServiceTest` test. This is an expected potential failure, and not related to the content of this PR. I'll re-invoke the tests shortly to give the Auth0 API some breathing room.

**Comment by MocanaAtBroad** [comment]
Patch did right thing for me, but I do get some weird error out stuff:
```
[builder] executing custom task: SingularAboutHealthySelectLogic
08:05:18.657   C: S: [main] INFO  o.b.d.s.t.SingularAboutHealthySelectLogic TASK:: SingularAboutHealthySelectLogic
08:05:18.959   C: S: [main] WARN  o.b.ddp.cache.CacheService Configuration file not set: jcacheConfigurationFileJCache is not enabled
JdbiPicklistOption.sql.stg 41:1: invalid character '<'
JdbiPicklistOption.sql.stg 41:1: invalid character '!'
JdbiPicklistOption.sql.stg 42:4: garbled template definition starting at 'The'
JdbiPicklistOption.sql.stg 42:8: garbled template definition starting at 'longer'
JdbiPicklistOption.sql.stg 42:15: garbled template definition starting at 'names'
JdbiPicklistOption.sql.stg 42:21: garbled template definition starting at 'here'
JdbiPicklistOption.sql.stg 42:26: garbled template definition starting at 'are'
JdbiPicklistOption.sql.stg 42:30: garbled template definition starting at 'due'
JdbiPicklistOption.sql.stg 42:34: garbled template definition starting at 'to'
JdbiPicklistOption.sql.stg 42:37: garbled template definition starting at 'the'
JdbiPicklistOption.sql.stg 42:41: garbled template definition starting at 'PicklistOptionDto'
JdbiPicklistOption.sql.stg 43:4: garbled template definition starting at 'not'
JdbiPicklistOption.sql.stg 43:8: garbled template definition starting at 'being'
JdbiPicklistOption.sql.stg 43:14: garbled template definition starting at 'completely'
JdbiPicklistOption.sql.stg 43:25: garbled template definition starting at 'Java'
JdbiPicklistOption.sql.stg 43:30: garbled template definition starting at 'Bean'
JdbiPicklistOption.sql.stg 43:35: garbled template definition starting at 'compatible'
JdbiPicklistOption.sql.stg 44:14: invalid character '-'
JdbiPicklistOption.sql.stg 44:16: invalid character '2'
JdbiPicklistOption.sql.stg 44:17: invalid character '0'
JdbiPicklistOption.sql.stg 44:18: invalid character '2'
JdbiPicklistOption.sql.stg 44:19: invalid character '2'
JdbiPicklistOption.sql.stg 44:5: garbled template definition starting at 'bskinner'
JdbiPicklistOption.sql.stg 44:21: invalid character '0'
JdbiPicklistOption.sql.stg 44:22: invalid character '8'
JdbiPicklistOption.sql.stg 44:24: invalid character '2'
JdbiPicklistOption.sql.stg 44:25: invalid character '9'
JdbiPicklistOption.sql.stg 45:0: invalid character '!'
JdbiPicklistOption.sql.stg 45:1: invalid character '>'
08:05:19.207   C: S: [main] INFO  o.b.d.s.t.SingularAboutHealthySelectLogic Successfully set isExclusive = true for [question:IF_YOU_HAVE_HAD_ARRHYTHMIA_COMPLICATIONS,option:DO_NOT_REQUIRE]
08:05:19.372   C: S: [main] INFO  o.b.d.s.t.SingularAboutHealthySelectLogic Successfully set isExclusive = true for [question:IF_YOU_HAVE_HAD_ARRHYTHMIA,option:DID_NOT_REQUIRED]
08:05:19.372   C: S: [main] INFO  o.b.d.s.t.SingularAboutHealthySelectLogic TASK:: SingularAboutHealthySelectLogic was successfully applied
[builder] done
```

**Comment by bskinner** [comment]
> Patch did right thing for me, but I do get some weird error out stuff:
> 
> ```
> [builder] executing custom task: SingularAboutHealthySelectLogic
> 08:05:18.657   C: S: [main] INFO  o.b.d.s.t.SingularAboutHealthySelectLogic TASK:: SingularAboutHealthySelectLogic
> 08:05:18.959   C: S: [main] WARN  o.b.ddp.cache.CacheService Configuration file not set: jcacheConfigurationFileJCache is not enabled
> JdbiPicklistOption.sql.stg 41:1: invalid character '<'
> JdbiPicklistOption.sql.stg 41:1: invalid character '!'
> JdbiPicklistOption.sql.stg 42:4: garbled template definition starting at 'The'
> JdbiPicklistOption.sql.stg 42:8: garbled template definition starting at 'longer'
> JdbiPicklistOption.sql.stg 42:15: garbled template definition starting at 'names'
> JdbiPicklistOption.sql.stg 42:21: garbled template definition starting at 'here'
> JdbiPicklistOption.sql.stg 42:26: garbled template definition starting at 'are'
> JdbiPicklistOption.sql.stg 42:30: garbled template definition starting at 'due'
> JdbiPicklistOption.sql.stg 42:34: garbled template definition starting at 'to'
> JdbiPicklistOption.sql.stg 42:37: garbled template definition starting at 'the'
> JdbiPicklistOption.sql.stg 42:41: garbled template definition starting at 'PicklistOptionDto'
> JdbiPicklistOption.sql.stg 43:4: garbled template definition starting at 'not'
> JdbiPicklistOption.sql.stg 43:8: garbled template definition starting at 'being'
> JdbiPicklistOption.sql.stg 43:14: garbled template definition starting at 'completely'
> JdbiPicklistOption.sql.stg 43:25: garbled template definition starting at 'Java'
> JdbiPicklistOption.sql.stg 43:30: garbled template definition starting at 'Bean'
> JdbiPicklistOption.sql.stg 43:35: garbled template definition starting at 'compatible'
> JdbiPicklistOption.sql.stg 44:14: invalid character '-'
> JdbiPicklistOption.sql.stg 44:16: invalid character '2'
> JdbiPicklistOption.sql.stg 44:17: invalid character '0'
> JdbiPicklistOption.sql.stg 44:18: invalid character '2'
> JdbiPicklistOption.sql.stg 44:19: invalid character '2'
> JdbiPicklistOption.sql.stg 44:5: garbled template definition starting at 'bskinner'
> JdbiPicklistOption.sql.stg 44:21: invalid character '0'
> JdbiPicklistOption.sql.stg 44:22: invalid character '8'
> JdbiPicklistOption.sql.stg 44:24: invalid character '2'
> JdbiPicklistOption.sql.stg 44:25: invalid character '9'
> JdbiPicklistOption.sql.stg 45:0: invalid character '!'
> JdbiPicklistOption.sql.stg 45:1: invalid character '>'
> 08:05:19.207   C: S: [main] INFO  o.b.d.s.t.SingularAboutHealthySelectLogic Successfully set isExclusive = true for [question:IF_YOU_HAVE_HAD_ARRHYTHMIA_COMPLICATIONS,option:DO_NOT_REQUIRE]
> 08:05:19.372   C: S: [main] INFO  o.b.d.s.t.SingularAboutHealthySelectLogic Successfully set isExclusive = true for [question:IF_YOU_HAVE_HAD_ARRHYTHMIA,option:DID_NOT_REQUIRED]
> 08:05:19.372   C: S: [main] INFO  o.b.d.s.t.SingularAboutHealthySelectLogic TASK:: SingularAboutHealthySelectLogic was successfully applied
> [builder] done
> ```

That's really strange. I'm not seeing that error and I'm not sure how to go about replicating it as `<! ... !>` is the StringTemplate comment construct- that block should be completely ignored. Checked my dependency tree on this branch (`ddp-8576-about-me-select-logic`) and maven resolves `org.jdbi:jdbi3-stringtemplate4:jar:3.28.0`. What commits are you using for building & running the patch?

Edit: It looks like `//` is being used elsewhere in the files for comments, but StringTemplate doesn't explicitly call that out as the structure for comments (although it is used in some of their examples). I'll update the file with the necessary changes.

**Review by MocanaAtBroad — APPROVED** [review]

**Pull request merged**